text/0003-data-frames.md

```diff
@@ -0,0 +1,252 @@
+
+To accomplish this streaming, we also introduce an `EndFrame`. Frame and EndFrame work together to allow a frame to be streamed as a multi-part frame, ending once the corresponding EndFrame has been read.
+
+As an example, let's say we were processing some content and wanted to output the first row and later the second row of this table:
+
+| tag  | length |
+| ---- | ------ |
+| head | 1024   |
+| body | 8192   |
+
+To do this, we would send the two separate frames, both marked as partial frames.
+
+```rust
+let frame_id = Uuid::new_v4();
+output.send(UntaggedValue::DataFrame {
+    headers: Some(vec!["tag", "length"]),
+    length: vec![
+        vec![ Value::from("head"), Value::from(1024)]
+    ],
+    partial_frame_id: Some(frame_id),
+    is_object_false,
+}.into_value());
+
+// ... time passes
+
+output.send(UntaggedValue::DataFrame {
+    headers: Some(vec!["tag", "length"]),
+    length: vec![
+        vec![ Value::from("body"), Value::from(8192)]
+    ],
+    partial_frame_id: Some(frame_id),
+    is_object_false,
+}.into_value());
+
+output.send(UntaggedValue::EndFrame(frame_id).into_value());
+```
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+Much of the implementation is part of the explanation in the previous section. In this section, we'll explore more of the impact of making this change.
+
+`UntaggedValue` will have `Table` and `Row` replaced with `DataFrame` (possibly just called `Frame` for brevity) and `EndFrame`.
+
+All commands that operate on Row today will need to be updated to work with the data frame instead.
+
+Rather than operating on a single row, commands will need to be updated to handle a frame at a time. Here, the mapping should largely be the same, though an additional inner loop to process over the rows will be necessary. This processing can be done serial or in parallel and may be done synchronously or asynchronously.
+
+Commands that worked over inner tables should be able to migrate to data frames, as there is a strong overlap in functionality.
+
+Commands that filter will work similarly as before, and may opt to output streams which are flattened by the output stream. This allows them to optionally return no frame if no rows in the frame met the requirements of the filter.
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+Some drawbacks come to mind:
+
+- This is a large, non-trivial amount of work. Getting this landed, updating the commands to use the new model, and thoroughly testing will take time.
+- This will break most, if not all, third-party plugins
+
+# Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Everything is a frame
+
+One alternative is to require everything to live inside of a frame. There are some advantages here: this is seemingly more uniform, but at the risk of overloading the data frame concept.
+
+# Prior art
```

Comment:
Since streaming seems to be a big motivator for this, I wonder if there's other prior art regarding streams.

Reply:
As a total outsider, I'd take a look at [Apache Arrow](https://arrow.apache.org/) here. A lot of their messaging/docs are focused on efficient columnar storage (which I assume is not relevant here), but they have two features that are probably interesting for Nu to learn from:

- Good support for "nested" types (e.g. arbitrary JSON, or ragged arrays): see https://arrow.apache.org/docs/python/data.html#type-metadata and https://arrow.apache.org/docs/format/Columnar.html
- A streaming format (although it's tuned towards RPC): see https://arrow.apache.org/docs/python/ipc.html and https://arrow.apache.org/docs/format/Flight.html

Reply:
High Level API Docs on Apache Arrow for Rust...

https://docs.rs/arrow/1.0.1/arrow/

Seems to have most of the relevant stuff needed
for generating ideas on how to move forward...

Reply:
https://github.com/nevi-me/rust-dataframe/blob/master/notes/update-01__04-04-2020.md

Some more thoughts on dataframes in rust using arrow and a dataframe package